File: apps/patter-voice-sdk/libraries/python/getpatter/dashboard/export.py
"""Data export utilities for the dashboard."""

from __future__ import annotations

import csv
import io
import json
from typing import Any


def calls_to_csv(calls: list[dict[str, Any]]) -> str:
    """Convert call list to CSV string.

    Columns: call_id, caller, callee, direction, started_at, ended_at,
    duration_s, cost_total, cost_stt, cost_tts, cost_llm, cost_telephony,
    avg_latency_ms, turns_count, provider_mode.
    """
    output = io.StringIO()
    writer = csv.writer(output)
    writer.writerow([
        "call_id", "caller", "callee", "direction", "started_at", "ended_at",
        "duration_s", "cost_total", "cost_stt", "cost_tts", "cost_llm",
        "cost_telephony", "avg_latency_ms", "turns_count", "provider_mode",
    ])

    for call in calls:
        metrics = call.get("metrics") or {}
        cost = metrics.get("cost", {})
        latency_avg = metrics.get("latency_avg", {})
        turns = metrics.get("turns", ())

        writer.writerow([
            call.get("call_id", ""),
            call.get("caller", ""),
            call.get("callee", ""),
            call.get("direction", ""),
            call.get("started_at", ""),
            call.get("ended_at", ""),
            metrics.get("duration_seconds", ""),
            cost.get("total", ""),
            cost.get("stt", ""),
            cost.get("tts", ""),
            cost.get("llm", ""),
            cost.get("telephony", ""),
            latency_avg.get("total_ms", ""),
            len(turns) if isinstance(turns, (list, tuple)) else "",
            metrics.get("provider_mode", ""),
        ])

    return output.getvalue()


def calls_to_json(calls: list[dict[str, Any]]) -> str:
    """Convert call list to JSON string."""
    return json.dumps(calls, default=str)

File: apps/patter-voice-sdk/libraries/typescript/src/dashboard/export.ts
/**
 * Data export utilities for the dashboard.
 */

import type { CallRecord } from './store';

/**
 * Convert call list to CSV string.
 *
 * Columns: call_id, caller, callee, direction, started_at, ended_at,
 * duration_s, cost_total, cost_stt, cost_tts, cost_llm, cost_telephony,
 * avg_latency_ms, turns_count, provider_mode
 */
export function callsToCsv(calls: CallRecord[]): string {
  const header = [
    'call_id', 'caller', 'callee', 'direction', 'started_at', 'ended_at',
    'duration_s', 'cost_total', 'cost_stt', 'cost_tts', 'cost_llm',
    'cost_telephony', 'avg_latency_ms', 'turns_count', 'provider_mode',
  ];

  const rows = [header.join(',')];

  for (const call of calls) {
    const m = (call.metrics || {}) as Record<string, unknown>;
    const cost = (m.cost as Record<string, number>) || {};
    const latencyAvg = (m.latency_avg as Record<string, number>) || {};
    const turns = m.turns;
    const turnsCount = Array.isArray(turns) ? turns.length : '';

    const row = [
      csvEscape(call.call_id || ''),
      csvEscape(call.caller || ''),
      csvEscape(call.callee || ''),
      csvEscape(call.direction || ''),
      call.started_at ?? '',
      call.ended_at ?? '',
      m.duration_seconds ?? '',
      cost.total ?? '',
      cost.stt ?? '',
      cost.tts ?? '',
      cost.llm ?? '',
      cost.telephony ?? '',
      latencyAvg.total_ms ?? '',
      turnsCount,
      m.provider_mode ?? '',
    ];

    rows.push(row.map(String).join(','));
  }

  return rows.join('\n') + '\n';
}

/**
 * Convert call list to JSON string.
 */
export function callsToJson(calls: CallRecord[]): string {
  return JSON.stringify(calls);
}

function csvEscape(value: string): string {
  if (value.includes(',') || value.includes('"') || value.includes('\n')) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}
